File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.java
class Solution {
    public List<List<Integer>> pacificAtlantic(int[][] heights) {
        if(heights==null || heights.length==0 || heights[0].length==0){
            return new ArrayList<>();
        }
        int rows = heights.length;
        int cols = heights[0].length;

        // Create visited arrays to track cells reachable by Pacific and Atlantic Oceans
        boolean[][] pacific = new boolean[rows][cols];
        boolean[][] atlantic = new boolean[rows][cols];

        performBFSForOcean(heights, pacific, getPacificBorderCells(rows, cols));
        performBFSForOcean(heights, atlantic, getAtlanticBorderCells(rows, cols));

        // Find and return the intersection of cells reachable by both oceans
        return findIntersectingCells(pacific, atlantic, rows, cols);
    }

    private List<Pair> getPacificBorderCells(int row, int col){
        List<Pair> bordercells = new ArrayList<>();
        for(int i=0; i<row; i++){
            bordercells.add(new Pair(i, 0)); //left colums
        }
        for(int j=0; j<col; j++){
            bordercells.add(new Pair(0, j)); // top row
        }
        return bordercells;
    }

    private List<Pair> getAtlanticBorderCells(int row, int col){
        List<Pair> bordercells = new ArrayList<>();
        for(int i=0; i<row; i++){
            bordercells.add(new Pair(i, col-1));
        }
        for(int j=0; j<col; j++){
            bordercells.add(new Pair(row-1, j));
        }
        return bordercells;
    }

    private void performBFSForOcean(int[][] heights, boolean[][] visited, List<Pair> bordercells){
        int rows = heights.length;
        int cols = heights[0].length;

        int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

        Queue<Pair> qu = new LinkedList<>(bordercells);

        for(Pair cell: bordercells){
            visited[cell.x][cell.y] = true;
        }

        //perfoem BFS
        while(!qu.isEmpty()){
            Pair cell = qu.poll();
            int x = cell.x;
            int y = cell.y;

            for (int[] dir : directions) {
                int nx = x + dir[0];
                int ny = y + dir[1];

                // Check bounds, height condition, and if the cell is already visited
                if (nx >= 0 && nx < rows && ny >= 0 && ny < cols
                        && !visited[nx][ny] && heights[nx][ny] >= heights[x][y]) {
                    visited[nx][ny] = true;
                    qu.offer(new Pair(nx, ny));
                }
            }
        }
    }

    private List<List<Integer>> findIntersectingCells(boolean[][] pacific, boolean[][] atlantic, int rows, int cols){
        List<List<Integer>> result = new ArrayList<>();
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                if (pacific[i][j] && atlantic[i][j]) {
                    result.add(Arrays.asList(i, j));
                }
            }
        }
        return result;
    }

}
class Pair{
    int x, y;

    public Pair(int x, int y){
        this.x = x;
        this.y = y;
    }
}

File: 0417-pacific-atlantic-water-flow/0417-pacific-atlantic-water-flow.py
from typing import List

class Solution:
    def pacificAtlantic(self, heights: List[List[int]]) -> List[List[int]]:
        if not heights or not heights[0]:
            return []

        rows, cols = len(heights), len(heights[0])
        pacific, atlantic = set(), set()

        def dfs(r, c, visited, prev_height):
            if (
                (r, c) in visited or
                r < 0 or c < 0 or r >= rows or c >= cols or
                heights[r][c] < prev_height
            ):
                return
            visited.add((r, c))
            directions = [(0, 1), (0, -1), (1, 0), (-1, 0)]
            for dr, dc in directions:
                dfs(r + dr, c + dc, visited, heights[r][c])

        # Start DFS from Pacific (top row and left column)
        for c in range(cols):
            dfs(0, c, pacific, heights[0][c])  # Top row
            dfs(rows - 1, c, atlantic, heights[rows - 1][c])  # Bottom row

        for r in range(rows):
            dfs(r, 0, pacific, heights[r][0])  # Left column
            dfs(r, cols - 1, atlantic, heights[r][cols - 1])  # Right column

        # Find the intersection of both ocean-reachable cells
        return list(pacific & atlantic)